### DISCUSSION TITLE
Some help with variable auto-completion

### DISCUSSION THREAD

**Discussion by paulfrench** [eclipse-langium/langium#353] [General]
Here is a simple grammar...

```
grammar Cleese
hidden(WS, SL_COMMENT, ML_COMMENT)

Model:
	(entities+=Entity)*;

Entity:
	'entity' name=ID ('extends' superType = [Entity])? '{'
		(properties+=Property)*
    ('keys' ':' '[' keys+=[Property]("," keys+=[Property])* ']' )?
	'}';

Property:
	(name=ID ':' type=PrimitiveType)*;

PrimitiveType returns string:
	"integer" | "string" |  "boolean";

terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;

terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

Here is a simple model based on this grammar

```
entity Person {
  id: integer
  firstname: string
  keys : [ id, firstname ]
}

entity Customer extends Person {
  surname: string
  keys : [surname]
}
```

For the **keys** property I cannot get auto-completion to work. It works no problem for adding an "extends" clause, you get a nice list of Entities you can reference. For **keys** property I just get the last property specified in the entity as a possible choice. For example **id** in the Person entity gives an error. Can I achieve this or do I need to dig deeper into scoping etc and the index file used.

Thanks

**Comment by msujew** [comment]
Hi @paulfrench,

That's weird, I just tried your grammar and there actually seems to be something wrong with the indexing. For example, when using only `firstname` here, the linker seems to link to the `id` property for some reason:

![image](https://user-images.githubusercontent.com/4377073/149144729-0f219c74-bcee-4c76-8ad9-5461727ad550.png)

I will investigate this one in a separate issue.

**Reply by paulfrench** [reply]
many thanks for taking a look. I'm using 0.2.0 if that helps in anyway. Cheers

**Comment by paulfrench** [comment]
Hello @msujew , I've not seen any issues created for this? Is it an issue?

**Comment by paulfrench** [comment]
Hello @msujew , not sure how this happened but there is a mistake in the grammar. Apologies for wasting your time...

The following...
```
Property:
	(name=ID ':' type=PrimitiveType)*;
```
...should obviously be
```
Property:
	name=ID ':' type=PrimitiveType
```
Best regards

**Reply by msujew** [reply]
Oh that's funny :D It's cool, I didn't even catch that one.

I assume once this is fixed, the code completion/cross reference resolving is working correctly?

**Comment by paulfrench** [comment]
Yes, it almost works like a charm. I still have one weird issue which I'm double checking the grammar for, for example...

```
entity Person {
  id: integer
  surname: string
  firstname: string
  keys : []
}
```
...if I place the cursor between the [] brackets then I get no sensible auto completion. I have to create a space first within the square brackets and then I get "id, surname and firstname" as possible choices.

**Reply by msujew** [reply]
Yeah, that's probably an issue with the code completion itself. It was the first LSP service which we implemented and it shows in its lack of quality...

I intend to redesign that part probably, for the 0.4.0 version (0.3 is already planned for February). The current implementation is doing some weird interpretation of the grammar, which does not work in all cases.

**Reply by paulfrench** [reply]
No worries I'll live with it for now
